### ISSUE TITLE
AI

### ISSUE THREAD

**Issue by doomsdayreturns** [ServUO/ServUO#5015]
issues with ai acquiring players if another mobile is in the path and they move through them they cannot step on that tile so they move in a circle until able to step on that tile.

![AI](https://user-images.githubusercontent.com/23180473/176102201-947643a3-650e-45b3-a11c-e7b0328f1fe4.jpg)

**Comment by whoblitzell** [comment]
I run a small online game based on this code repository, a shard of a game from the UO era that was VERY similar to UO in almost all respects. Anyways, the way I addressed this problem was to use the "Diagonal" method described below:

https://theory.stanford.edu/~amitp/GameProgramming/Heuristics.html

This can be easily integrated by making methods such as CanFit() and CanPlayerFit() on Map.cs and then having a MapPath.cs that can help a Pathfinding module use A-star algorithm or others to compute legal paths to player. Rather than moving in a circle, the algorithm will regard the non-goal player as an obstacle and simply move around them as they would anything else. An alternative method and an easier, cheaper one is to allow the mobile doll to occupy the same space as the opposing mobile temporarily on the server, although that can create client side issues for many games!

I would advise anyone doing this to simply extend the code as described above.

**Comment by kamronbatman** [comment]
It should already do what you describe. I wonder if this happens with the fixed A* changes on ModernUO. Should test it and see.
The bug we encountered from RunUO (and presumably ServUO has it too), is that in some scenarios, the mob will try to go around in one way, and then try another way, with no "give up" scenario.

**Comment by whoblitzell** [comment]
Yeah late to reply, but we ended up with an IsOccupied check in Map.cs that worked off Map.GetMobileInRange() and Map.GetPlayersInRange(), checking the needed properties to ignore mobile types you want to ignore as obstacles.

**Comment by kamronbatman** [comment]
I dived into this more, but finding an elegant and generalized solution is still quite a bit of work. I think the isoccupied can work for small shards. The complexity comes when you try to mark off a tile, but the mobile moves and you have to effectively repath.

What I haven't dived into is a pathing algorithm that easily facilitates dynamic rerouting. Probably caching paths ahead of time and adjusting accordingly would help significantly.

If anyone has any ideas on an approach please let us know!

**Comment by whoblitzell** [comment]
Yeah we pre calculate the paths and adjust in the fly. Invalidate cache
based on world  clock

**Comment by whoblitzell** [comment]
Ok so why not use is occupied check and cache it and with the neighbors
nodes in a dictionary right. When player moves out of a tile remove that
entry from cache and give it an absolute cache expiry of like a minute.

I used vanilla a star in my servuo with this approach and it works fine. I
no longer encounter the issue you described

**Comment by whoblitzell** [comment]
As for give up logic..  keep a hash set of visited nodes and only process
each node once per path request. Also add a Max loop count for rather max
count for processed nodes... When giving up return a partial path. Wait for
cache expiration and future path request work when the node isn't occupied
anymore

**Comment by whoblitzell** [comment]
What strikes me about SlowAStarAlgorithm is a few things. The first thing is open and closed lists should probably be ArrayList<PathNode> (arrays need resizing too often) and thus not existing as PathNode[]. Backtracking should work off a separately stored path not just the close list (what if node is reset or not there anymore? We fail and return null...)

I'd wonder why to not just use Manhattan or Diagonal heuristic rather than squared, but thats just a gripe. Nodes stored in open and closed lists need to be properly initialized (new PathNode(...) custom constructor) so that their state is valid.

The main loop needs terminating conditions. For example if depth >= maxDepth, it should return a partial path and/or null. A HashSet or Dictionary of processed nodes should be stored (locally or at class level) as to not process duplicate nodes (continue;)

Rather than this:

              if (openCount == MaxNodes)
                        break;

and this:

                if (openCount == MaxNodes || closedCount == MaxNodes)
                    break;

This instead:
if (openCount == 0)
{
    break;
} 

and:

if (openCount >= MAX_PROCESSED_NODE_COUNT || closedCount >= MAX_PROCESSED_NODE_COUNT)
    break;
[SlowAStarAlgorithm_Fixed.txt](https://github.com/user-attachments/files/17093046/SlowAStarAlgorithm_Fixed.txt)

**Comment by kamronbatman** [comment]
I believe the pathing is recreated on every move request, therefore a state machine would have to be introduced to the NPCs AI so there is an existing partial path and/or cache to draw from.

**Comment by whoblitzell** [comment]
Guess I don't follow. Why does it need to persist past pathing?
